api/src/services/schema.ts

```diff
@@ -0,0 +1,290 @@
+					const existingRelation = snapshotWithHash.relations.find(
+						(r) => r.collection === diffRelation.collection && r.field === diffRelation.field
+					);
+					if (!existingRelation) {
+						throw new InvalidPayloadException(
+							`Provided diff is trying to delete relation "${relation}" but it does not exist. Please regenerate a new diff and try again.`
+						);
+					}
+				} else {
+					const matchingRelation = snapshotWithHash.relations.find((r) => r.hash === diffRelation.hash);
+					if (!matchingRelation) {
+						throw new InvalidPayloadException(
+							`Provided diff for relation "${relation}" does not match the current instance's relation, indicating it has changed after this diff was generated. Please regenerate a new diff and try again.`
+						);
+					}
+				}
+			}
+
+			throw new InvalidPayloadException(
+				`Provided hash "${payload.hash}" does not match the current instance's schema hash "${snapshotWithHash.hash}". Please regenerate a new diff and try again.`
+			);
+		}
+
+		if (
+			payload.diff.collections.length === 0 &&
+			payload.diff.fields.length === 0 &&
+			payload.diff.relations.length === 0
+		) {
+			return;
+		}
+
+		await applyDiff(currentSnapshot, payload.diff, { database: this.knex });
+	}
+
+	async diff(
+		snapshot: Snapshot,
+		options?: { currentSnapshot?: Snapshot; force?: boolean }
+	): Promise<SnapshotDiff | null> {
+		if (this.accountability?.admin !== true) throw new ForbiddenException();
+
+		if (!options?.force) {
+			if (snapshot.directus !== currentDirectusVersion) {
+				throw new InvalidPayloadException(
+					`Provided snapshot's directus version ${snapshot.directus} does not match the current instance's version ${currentDirectusVersion}. You can bypass this check by passing the "force" query parameter.`
+				);
+			}
+
+			if (!snapshot.vendor) {
+				throw new InvalidPayloadException(
+					'Provided snapshot does not contain the "vendor" property. You can bypass this check by passing the "force" query parameter.'
+				);
+			}
+
+			const currentVendor = getDatabaseClient();
+			if (snapshot.vendor !== currentVendor) {
+				throw new InvalidPayloadException(
+					`Provided snapshot's vendor ${snapshot.vendor} does not match the current instance's vendor ${currentVendor}. You can bypass this check by passing the "force" query parameter.`
+				);
+			}
+		}
+
+		const { error } = snapshotJoiSchema.validate(snapshot);
+		if (error) throw new InvalidPayloadException(error.message);
+
+		const currentSnapshot = options?.currentSnapshot ?? (await getSnapshot({ database: this.knex }));
+		const diff = getSnapshotDiff(currentSnapshot, snapshot);
+
+		if (diff.collections.length === 0 && diff.fields.length === 0 && diff.relations.length === 0) {
+			return null;
+		}
+
+		return diff;
+	}
+
+	getHashedSnapshot(snapshot: Snapshot): SnapshotWithHash {
+		const snapshotHash = getVersionedHash(snapshot);
+
+		return {
+			hash: snapshotHash,
+			...snapshot,
```

Comment:
```suggestion
			...snapshot,
			hash: snapshotHash,
```

Reply:
I believe the intent (at the time) was to simply follow the "hash at the front" logic for the collections/fields/relations:

https://github.com/directus/directus/blob/776c84c2634b03010ed6e4663876b98f583d92d9/api/src/services/schema.ts#L295-L300

but in hindsight I do think there isn't an actual purpose in keeping it in front at this point!